File: backend/app/schemas/plans.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field


class FilmRequestLineCreate(BaseModel):
    material: str
    color: str
    thickness: float
    planned_area_m2: float = Field(gt=0)


class FilmRequestLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    material: str
    color: str
    thickness: float
    planned_area_m2: float
    current_stock_m2: float
    shortage: bool


class WeeklyPlanCreate(BaseModel):
    week_start: date
    week_end: date


class WeeklyPlanOut(BaseModel):
    id: int
    week_start: date
    week_end: date
    created_by: int
    status: str
    lines: list[FilmRequestLineOut]


class PlanFactLine(BaseModel):
    line_id: int
    material: str
    color: str
    thickness: float
    planned_area_m2: float
    actual_area_m2: float
    percent_complete: float
    by_width: dict[float, float]


class PlanFactOut(BaseModel):
    week_id: int
    week_start: date
    week_end: date
    lines: list[PlanFactLine]

File: frontend/src/api/plans.ts
import { apiClient } from "./client";

export interface FilmRequestLine {
  id: number;
  material: string;
  color: string;
  thickness: number;
  planned_area_m2: number;
  current_stock_m2: number;
  shortage: boolean;
}

export interface WeeklyPlan {
  id: number;
  week_start: string;
  week_end: string;
  created_by: number;
  status: string;
  lines: FilmRequestLine[];
}

export interface FilmRequestLineCreate {
  material: string;
  color: string;
  thickness: number;
  planned_area_m2: number;
}

export interface PlanFactLine {
  line_id: number;
  material: string;
  color: string;
  thickness: number;
  planned_area_m2: number;
  actual_area_m2: number;
  percent_complete: number;
  by_width: Record<string, number>;
}

export interface PlanFactOut {
  week_id: number;
  week_start: string;
  week_end: string;
  lines: PlanFactLine[];
}

export async function createWeeklyPlan(payload: { week_start: string; week_end: string }): Promise<WeeklyPlan> {
  const { data } = await apiClient.post<WeeklyPlan>("/weekly-plans", payload);
  return data;
}

export async function listWeeklyPlans(): Promise<WeeklyPlan[]> {
  const { data } = await apiClient.get<WeeklyPlan[]>("/weekly-plans");
  return data;
}

export async function getWeeklyPlan(id: number): Promise<WeeklyPlan> {
  const { data } = await apiClient.get<WeeklyPlan>(`/weekly-plans/${id}`);
  return data;
}

export async function addFilmRequestLine(planId: number, payload: FilmRequestLineCreate): Promise<FilmRequestLine> {
  const { data } = await apiClient.post<FilmRequestLine>(`/weekly-plans/${planId}/lines`, payload);
  return data;
}

export async function getPlanFact(weekId: number): Promise<PlanFactOut> {
  const { data } = await apiClient.get<PlanFactOut>("/plan-fact", { params: { week_id: weekId } });
  return data;
}
